File: frontend/src/types/resume.ts
export interface ResumeListItem {
  id: string
  name: string
  template_id: string
  updated_at: string
}

export interface Resume {
  version: string
  meta: ResumeMeta
  personal: Personal
  /**
   * Standalone "个人总结" (personal summary) block. Wrapped in its own struct so a
   * single Hidden flag can toggle the whole block while keeping the text on disk.
   * Mirrors the Go `PersonalSummary` struct (json: personal_summary).
   */
  personal_summary?: PersonalSummary
  internships?: Internship[]
  jobs?: Job[]
  projects?: Project[]
  education?: Education[]
  skills?: SkillGroup[]
  languages?: Language[]
  awards?: Award[]
  custom?: CustomSection[]
  /**
   * Per-resume custom CSS（样式定制单一承载体，见 lib/customCss.ts）。
   * 承载页边距/内容间距/头像尺寸/圆角/信息区布局等全部样式调整；为空表示
   * 无样式定制，渲染时呈现模板原生外观。数据模型不再为单个样式特性新增字段。
   */
  custom_css?: string
}

/**
 * Base font size levels (pt). The wire format stays numeric so resumes
 * persisted before this enum was introduced remain valid; assignments
 * must go through these constants (mirrors model.FontSize in Go).
 */
export const FONT_SIZE_LEVELS = {
  small: 9,
  medium: 10,
  large: 11,
} as const

export type FontSizeKey = keyof typeof FONT_SIZE_LEVELS

export const DEFAULT_FONT_SIZE: number = FONT_SIZE_LEVELS.medium

export interface ResumeMeta {
  template_id: string
  language: string
  /** Base font size in pt; assign via FONT_SIZE_LEVELS (see above). */
  font_size: number
  created_at: string
  updated_at: string
  export_count: number
  name: string
}

export interface Personal {
  full_name: string
  english_name?: string
  email?: string
  phone?: string
  wechat?: string
  qq?: string
  location?: string
  website?: string
  linkedin?: string
  github?: string
  avatar?: string
  birthday?: string
  gender?: 'male' | 'female' | 'other'
  job_title?: string
  years_of_exp?: number
}

export interface PersonalSummary {
  summary?: string
  /**
   * Optional visibility toggle for the "个人总结" block. When true the summary
   * text is dropped from the rendered resume while kept on disk. Mirrors the Go
   * struct's `Hidden *bool` (nil = visible, for legacy-data compatibility).
   */
  hidden?: boolean
}

export interface Internship {
  id: string
  company: string
  company_url?: string
  title: string
  location?: string
  start_date: string
  end_date?: string
  is_current: boolean
  summary?: string
  highlights?: string[]
  keywords?: string[]
  /**
   * Optional visibility toggle. When true the entry is hidden from the rendered
   * resume but kept in the editor. Optional so legacy data remains valid.
   */
  hidden?: boolean
}

export interface Job {
  id: string
  company: string
  company_url?: string
  title: string
  location?: string
  start_date: string
  end_date?: string
  is_current: boolean
  summary?: string
  highlights?: string[]
  keywords?: string[]
  hidden?: boolean
}

export interface ExtraField {
  id: string
  label: string
  value: string
}

export interface Project {
  id: string
  name: string
  url?: string
  role?: string
  start_date?: string
  end_date?: string
  is_current?: boolean
  summary?: string
  highlights?: string[]
  keywords?: string[]
  extras?: ExtraField[]
  hidden?: boolean
}

export interface Education {
  id: string
  school: string
  degree: string
  major: string
  minor?: string
  start_date: string
  end_date: string
  gpa?: string
  courses?: string
  highlights?: string[]
  hidden?: boolean
}

export interface SkillGroup {
  id: string
  category: string
  items: Skill[]
  hidden?: boolean
}

export interface Skill {
  name: string
  level?: number
  icon?: string
  hidden?: boolean
}

export interface Language {
  id: string
  name: string
  level: string
  proficiency?: string
  hidden?: boolean
}

export interface Award {
  id: string
  title: string
  date: string
  issuer?: string
  summary?: string
  hidden?: boolean
}

export interface CustomSection {
  id: string
  title: string
  items: CustomItem[]
  hidden?: boolean
}

export interface CustomItem {
  id: string
  title: string
  subtitle?: string
  date?: string
  description?: string
  highlights?: string[]
  hidden?: boolean
}

export function createEmptyResume(templateId: string): Resume {
  return {
    version: '1.1',
    meta: {
      template_id: templateId,
      language: 'zh-CN',
      font_size: DEFAULT_FONT_SIZE,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
      export_count: 0,
      name: '',
    },
    personal: {
      full_name: '',
    },
    personal_summary: { summary: '' },
    internships: [],
    jobs: [],
    projects: [],
    education: [],
    skills: [],
    languages: [],
    awards: [],
    custom: [],
  }
}

export function generateId(): string {
  return crypto.randomUUID ? crypto.randomUUID() : Math.random().toString(36).slice(2, 10)
}

/**
 * Migrates legacy resume data where the standalone summary lived as top-level
 * `summary` / `summary_hidden` fields onto the `personal_summary` struct.
 * Newer data already using `personal_summary` is returned unchanged. Call this
 * right after loading a resume so the rest of the app can assume the new shape.
 */
export function migratePersonalSummary(resume: Resume): Resume {
  if (resume.personal_summary || (resume as unknown as Record<string, unknown>).summary === undefined) {
    return resume
  }
  const legacy = resume as unknown as { summary?: string; summary_hidden?: boolean }
  return {
    ...resume,
    personal_summary: {
      summary: legacy.summary ?? '',
      hidden: legacy.summary_hidden,
    },
    // strip the deprecated top-level fields
    summary: undefined,
    summary_hidden: undefined,
  } as Resume
}

File: pkg/resume/model/resume.go
package model

import (
	"encoding/json"
	"fmt"
	"time"
)

// Resume 是简历的根结构。
//
// SummaryHidden 控制「个人总结」区块是否显示。使用指针 + omitempty 是为了兼容
// 历史数据：字段缺失时视为显示。
type Resume struct {
	Version         string          `json:"version"`
	Meta            ResumeMeta      `json:"meta"`
	Personal        Personal        `json:"personal"`
	PersonalSummary PersonalSummary `json:"personal_summary"`
	Internships     []Internship    `json:"internships,omitempty"`
	Jobs            []Job           `json:"jobs,omitempty"`
	Projects        []Project       `json:"projects,omitempty"`
	Education       []Education     `json:"education,omitempty"`
	Skills          []SkillGroup    `json:"skills,omitempty"`
	Languages       []Language      `json:"languages,omitempty"`
	Awards          []Award         `json:"awards,omitempty"`
	Custom          []CustomSection `json:"custom,omitempty"`
	CustomCSS       string          `json:"custom_css,omitempty"`
}

type PersonalSummary struct {
	Summary string `json:"summary,omitempty"`
	Hidden  *bool  `json:"hidden,omitempty"`
}

// FontSize 是简历的基准字号，单位 pt。传输格式保持数字，以便该枚举引入前
// 已持久化的简历仍然有效。
type FontSize int

// 字号档位取值。
const (
	FontSizeSmall  FontSize = 9
	FontSizeMedium FontSize = 10
	FontSizeLarge  FontSize = 11
)

// ResumeMeta 保存简历文档的元数据。
type ResumeMeta struct {
	TemplateID  string    `json:"template_id"`
	Name        string    `json:"name"`
	Language    string    `json:"language"`
	FontSize    FontSize  `json:"font_size"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	ExportCount int       `json:"export_count"`
}

// SchemaVersion 是简历数据的 schema 版本，与 App 版本（config.yaml 的 app.version）
// 解耦。字段结构未变时保持 "1.0"。NewResume 与前端 sampleData 均以本常量为准。
//
// v1.1：新增 per-resume custom_css（样式定制统一由注入 CSS 承载），并从
// personal 移除 avatar_width/height/avatar_radius/header_layout 四个样式字段。
// 不再做样式迁移兼容：旧数据缺失这些样式时 custom_css 为空，渲染回退模板原生外观。
const SchemaVersion = "1.1"

// Migrate 把原始 JSON 反序列化为 Resume，并统一 version 为当前 SchemaVersion。
// 不迁移任何字段：本次改动仅涉及样式渲染，旧数据丢失的样式（页边距/内容间距/
// 头像尺寸/圆角/布局）由模板初始值兜底。
func Migrate(rawJSON []byte) (*Resume, error) {
	var versionInfo struct {
		Version string `json:"version"`
	}
	if err := json.Unmarshal(rawJSON, &versionInfo); err != nil {
		return nil, fmt.Errorf("parse version: %w", err)
	}

	switch versionInfo.Version {
	case SchemaVersion, "1.0", "1.0.0":
		var r Resume
		if err := json.Unmarshal(rawJSON, &r); err != nil {
			return nil, fmt.Errorf("unmarshal %q: %w", versionInfo.Version, err)
		}
		r.Version = SchemaVersion
		return &r, nil
	case "":
		return nil, fmt.Errorf("missing version field, cannot load")
	default:
		return nil, fmt.Errorf("unsupported data version: %s", versionInfo.Version)
	}
}
